tests/main/user-mounts/task.yaml

```diff
@@ -0,0 +1,40 @@
+summary: smoke tests for user mounts
+details: |
+    When a confined snap has a user-fstab file, additional bind mounts will
+    be performed prior to running the application.  These mounts are private
+    to the invocation of the application, so can be used to manipulate how
+    per-user data is presented.
+prepare: |
+    . $TESTSLIB/snaps.sh
+    install_local test-snapd-sh
+    rm -f /var/lib/snapd/mount/snap.test-snapd-sh.user-fstab
+execute: |
+    . $TESTSLIB/dirs.sh
+
+    echo "Create XDG_RUNTIME_DIR for test user"
+    USER_RUNTIME_DIR=/run/user/12345
+    mkdir -p $USER_RUNTIME_DIR || true
+
+    # Add some content to the runtime dir
+    rm -rf $USER_RUNTIME_DIR/snap.test-snapd-sh
+    mkdir $USER_RUNTIME_DIR/snap.test-snapd-sh
+    mkdir $USER_RUNTIME_DIR/snap.test-snapd-sh/source
+    mkdir $USER_RUNTIME_DIR/snap.test-snapd-sh/target
+    touch $USER_RUNTIME_DIR/snap.test-snapd-sh/source/in-source
+    touch $USER_RUNTIME_DIR/snap.test-snapd-sh/target/in-target
+
+    chmod -R u=rX,go= $USER_RUNTIME_DIR
+    chown -R test:test $USER_RUNTIME_DIR
+
+    echo "Without user-fstab we see target"
+    XDG_RUNTIME_DIR=$USER_RUNTIME_DIR su -p -l -c 'test-snapd-sh -c "ls \$XDG_RUNTIME_DIR/target"' test | MATCH in-target
+
+    echo "Configuring user-fstab"
+    cat << \EOF > /var/lib/snapd/mount/snap.test-snapd-sh.user-fstab
+    $XDG_RUNTIME_DIR/source $XDG_RUNTIME_DIR/target none bind,rw 0 0
```

Comment:
While I think that making the mountpoint mounting bulletproof is the ideal and what we strive for, depending on how that works it, it *may* be ok to try to detect if we mismounted (ie, lost the race) after the fact. I definitely prefer the former, but thought I put that out there.

Reply:
My main goal for user mounts is to support xdg-document-portal.  That involves creating a mount point from `$XDG_RUNTIME_DIR/doc/by-app/$app_id` to `$XDG_RUNTIME_DIR/doc`.  That source path is provided by a FUSE daemon running as the user, which I'd think would fail the above check, no?

Reply:
With both `$XDG_RUNTIME_DIR/doc/by-app/$app_id` and `$XDG_RUNTIME_DIR/doc`, both are under control of the calling user which because of CAP_SYS_ADMIN and the mount() via the setuid snap-confine calling the priv-dropped snap-update-ns is able to perform symlink races to have anything bind mounted anywhere (when snap-confine is not confined by apparmor), which is a privilege escalation (I'm not trying to protect the user from herself, but I do want to make sure that the user can't do more than what would normally be allowed to the user).

The directories you've outlined seem problematic in this regard unless @zyga has some thoughts regarding his "shadow mounts" technique or perhaps for using MS_MOVE (I've not played with either). 

I can say that mount will canonicalize both the source and target as they appear mountinfo, so it is certainly possible to verify after the bind mount call if these directories are not present (which would indicate someone injected a symlink), and if not umount, or even better snap-discard-ns. I've not thought this all through, but OTOH:

 1. drop privs
 2. verify that both source and target are in a single entry in mountinfo, else die()(or return 0?)
 3. perform the mount
 4. verify that both source and target are in a single entry in mountinfo, else discard and die()

Considering the above, I think we'd want to reconsider locking.

Reply:
@jhenstridge and I discussed whether locking was need for this (since I recommended reconsidering it), and he convinced me it isn't needed for this PR.

Reply:
I have some ideas on how to do this without any races (famous last words)